app/controllers/provider_interface/provider_agreements_controller.rb

```diff
@@ -12,11 +12,16 @@ def new_data_sharing_agreement
 
     def create_data_sharing_agreement
       @provider_agreement = ProviderAgreement.new(provider_agreement_params.merge(provider_user: current_provider_user))
-      if @provider_agreement.save
-        redirect_to provider_interface_path
-      else
-        render :data_sharing_agreement
-      end
+
+      render :data_sharing_agreement and return unless @provider_agreement.save
```

Comment:
Somehow I would find a nested `if` easier to follow here. But don't feel strongly about it. :)

```ruby
  if @provider_agreement.save
    @provider_setup = ProviderSetup.new(provider_user: current_provider_user)

    if @provider_setup.next_agreement_pending
      redirect_to provider_interface_new_data_sharing_agreement_path
    else
      @provider_relationship = @provider_setup.next_relationship_pending
      render :success
    end
  else
    render :data_sharing_agreement
  end
```

Reply:
I think maybe there's an issue with having 3 differing rendering/redirecting outcomes from a single action and that makes for a lot of conditions. I went with a more guard-clause-like format for brevity.